### PR TITLE
fix(pipeline): forward termination-zeroed discount to HordeActorCriticAgent in horde_ac mode

### LINKED ISSUE #2344: AlbertaPipeline control_mode="horde_ac" drops `terminated`: the value head bootstraps and the actor trace survives every episode boundary
`AlbertaPipeline.update` documents a `terminated` argument ("Scalar termination
flag (`0.0` or `1.0`)") and validates it as a float32 scalar restricted to
`{0, 1}`. On the default `control_mode="sarsa"` path it is forwarded into
`step4_update` → `SARSAAgent.update`, where it skips bootstrapping
(`skip_bootstrap = (safe_terminated != 0) | (gamma_arr == 0.0)`,
`alberta_framework/core/sarsa.py:945`) and zeroes the eligibility traces
(`sarsa.py:1005-1006`).

On the registered `control_mode="horde_ac"` path it is dropped on the floor.
`alberta_framework/pipeline.py:1576-1581` calls

```python
ac_result = ac.update(
    ac_state,
    reward,
    features,
    auxiliary_cumulants=auxiliary_cumulants,
)
```

with no `discount`, so `HordeActorCriticAgent.update` falls back to
`value_discount = self._critic.horde_spec.gammas[cfg.value_head_index]`
(`alberta_framework/core/horde_actor_critic.py:1001-1003`) on *every* step,
terminal or not.

Two things are consequently wrong at a terminal transition:

1. **The value head bootstraps through termination.** The critic's TD target
   for the value demon is `r + gamma * V(s')` instead of `r`.
2. **The actor eligibility trace survives the boundary.** In
   `HordeActorCriticAgent.update`, `actor_decay = value_discount * cfg.actor_lamda`
   and `carry_traces = value_discount != 0.0` (`horde_actor_critic.py:1044,1061`),
   so with `discount = 0` the actor trace is zeroed — but the pipeline never
   supplies it, and the trace is carried into the next episode instead.

The mechanism to fix it already exists and was built for exactly this caller.
`HordeActorCriticAgent.update` takes `discount: Array | None` — "Optional
scalar per-transition value-head discount" — and routes to
`HordeLearner.update_with_discounts`, whose docstring says it "lets control
adapters zero the value head at episode boundaries while keeping the Horde's
fixed GVF metadata and per-head trace decay intact"
(`alberta_framework/core/horde.py:554-557`). The pipeline is such a control
adapter and is the only in-tree one that does not use it.

## Reproduction

`alberta_framework/pipeline.py` at `c9aba7b5`, identity Step 2 features,
`horde.gammas=(0.9, 0.5)`, `value_head_index=0`, `actor_lamda=0.8`:

```python
state = pipeline.init(jr.key(0), jnp.asarray([0.2, -0.1, 0.4], dtype=jnp.float32))
obs = jnp.asarray([0.1, 0.3, -0.2], dtype=jnp.float32)
reward = jnp.asarray(0.5, dtype=jnp.float32)
cumulants = jnp.asarray([0.5, -0.2], dtype=jnp.float32)
for term in (0.0, 1.0):
    res = pipeline.update(state, obs, reward, jnp.asarray(term, dtype=jnp.float32), cumulants)
    print(term, res.horde_td_targets, res.state.control_state.actor_trace_weights)
```

```
0.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
1.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
```

The two lines are byte-identical: `terminated` has **no effect whatsoever** on
this control path. The value-head target should be `0.5` (the reward alone) in
the terminal case; it is `0.5 + 0.9 * V(s') = 0.5 + 0.9 * (-0.1718) = 0.3454`.
The actor trace should be empty; it is not.

## Why no test caught it

Every existing pipeline test passes `terminated = 0.0`
(`tests/test_pipeline.py:237`, `:280`, `:324`, `:412`, `:484`, `:540`, `:580`;
the array-run helper builds `terminated = jnp.zeros(n_steps)` at `:632`). `terminated = 1.0` is exercised nowhere in
the pipeline suite, on either control mode, so the entire termination semantics
of the documented flag are untested and the `horde_ac` branch's untested path
is silently wrong.

## Scope

`AlbertaPipeline.run_arrays` delegates to `AlbertaPipeline.update`, so array
runs and `run_pipeline_smoke` inherit the same defect. Any measurement taken
through the pipeline with `control_mode="horde_ac"` on an episodic stream
learns terminal transitions as if the episode continued. Auxiliary GVF demons
are *not* affected and must keep their own configured gammas — only the
value head's discount is a per-transition control quantity.

## Proposed fix

Pass the terminal-aware discount in the `horde_ac` branch:

```python
value_gamma = self._horde.horde_spec.gammas[value_index]
control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))
ac_result = ac.update(..., discount=control_discount)
```

At `terminated == 0` this is a bit-exact reduction: `update_with_discounts`
computes `bootstrap = jnp.where(discounts == 0.0, 0.0, discounts * next_preds)`
from the same `gammas` vector the `discount is None` branch uses, and the extra
finite/range head-validity conditions are satisfied by the configured gammas.

Lane/metric: this is a measurement-trustworthiness fix, not a benchmark climb —
no `outputs/` artifact changes and no registered evidence source is touched.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2344

## Summary
In `AlbertaPipeline.update`, when running under `control_mode="horde_ac"`, `terminated` was dropped when invoking `HordeActorCriticAgent.update`. Because `discount` was not passed, the agent fell back to the static `gammas[value_head_index]` on every step, causing:
1. The critic value head to bootstrap through episode termination (`r + gamma * V(s')` instead of `r`).
2. The actor eligibility trace to survive into the next episode instead of being zeroed at termination.

## Proposed Changes
1. Forwarded `control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))` to `ac.update(..., discount=control_discount)` in `AlbertaPipeline.update`.
2. Added unit test in `tests/test_pipeline.py` verifying that when `terminated=1.0`, the value-head target evaluates to `reward` alone (`0.5`) without bootstrapping.

## Verification
- Ran pytest on pipeline test suites: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_pipeline.py` (184/184 tests passed).
- Ran linter and type checker: `/opt/homebrew/bin/uv run ruff check alberta_framework/pipeline.py` & `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/pipeline.py` (all checks passed).

### CODE REVIEW

**Review by saltwire-dev — CHANGES_REQUESTED** [review]
## Request changes at exact head `6b0c0ebd520eba4c54b45a57b4b48a43e2bbe60e`

Reviewed at exact head `6b0c0ebd520eba4c54b45a57b4b48a43e2bbe60e` against production `alberta_framework/pipeline.py` and unchanged `horde_actor_critic.py` (not just the new test). Conversation sidebar is No reviews; the github-actions “produced no report” note is a comment, not a review. CPython 3.13.5, numpy 2.2.4.

Forwarding a terminal-zeroed `discount=` into `HordeActorCriticAgent.update` is the right *direction* for #2344 — origin drops `terminated` on the `horde_ac` branch, the agent falls back to the static value-head gamma, and both named defects follow: `r + γ V(s')` instead of `r`, and `carry_traces = value_discount != 0` leaves the actor eligibility trace live. The hole is that #2344 published **two** tensors on this exact recipe, and the new test only locks the first. The comment says “Traces in terminal step should not carry forward” and then asserts a TD-target inequality. `tests/test_pipeline.py:1201-1229` never reads `state.control_state.actor_trace_weights`. It never pins the issue’s continuing value-head target `0.34537804`. It never pins that auxiliary head 1 stays on its own γ=0.5. Sibling 1-commit #2528, same overlay, at least asserts `actor_trace_weights == 0`. Deleting the trace-zeroing half of the agent contract (or over-zeroing every Horde discount) would still pass this PR’s two asserts. On this CPU, a numpy replica of the **exact HAC actor-trace formulas at this SHA** reproduces the issue’s origin matrix `[[0.1, -0.05, 0.2], [-0.1, 0.05, -0.2]]` when `discount` is omitted, and zeros when `discount=0` — a measurement this test never takes.

**Blocking: named actor-trace half of #2344 is unwitnessed; continuing / aux tensors unpinned; comment asserts the wrong object**

On this CPU, a numpy replica of the **exact HAC formulas at this SHA** (unchanged from origin) with the issue’s last-action=0 / uniform-policy initial condition:

```
origin discount=None / 0.9  actor_trace_weights
  [[ 0.1  -0.05  0.2 ]
   [-0.1   0.05 -0.2 ]]     # byte-match of #2344's printed terminal traces
head   discount=0           actor_trace_weights
  [[ 0.    0.    0.  ]
   [ 0.    0.    0.  ]]
origin value-head target  0.5 + 0.9 * V(s') = 0.34537804
head   value-head target  0.5
```

So the production overlay *should* repair both named defects. The new test only proves `res_term.horde_td_targets[0] == 0.5` and that the two target *vectors* differ. That second assert is not an actor-trace check: auxiliary head 1 is supposed to stay on γ=0.5, so the inequality is just “head 0 moved.” It would still pass if a later change zeroed **every** Horde discount. It would still pass if `carry_traces` in the unchanged agent stopped following `value_discount` — the pipeline test never looks at `actor_trace_weights`.

#2344’s own print is the missing oracle: origin terminal traces are a specific nonzero 2×3 matrix. After a correct overlay they must be zeros **and** the continuing row must stay `[0.34537804, -0.18965168]`. Sibling 1-commit #2528 against the same recipe at least asserts `jnp.allclose(res_term.state.control_state.actor_trace_weights, 0.0)` and comments the `0.3454` bootstrap; this PR’s test dropped both. `run_arrays` still builds `terminated = jnp.zeros(n_steps)`.

**What I would approve**

Keep the `discount=` overlay. Lock **both** tensors #2344 printed, on the public `AlbertaPipeline.update` path, at this recipe. (1) Actor traces: after `pipeline.update(..., terminated=1.0)` from the issue’s `jr.key(0)` init, `res_term.state.control_state.actor_trace_weights` and `actor_trace_bias` are exact zeros. Origin on the same call is the published `[[0.1, -0.05, 0.2], [-0.1, 0.05, -0.2]]`. (2) Value-head targets: terminal head 0 is `0.5`; continuing head 0 is the published `0.34537804` (not merely “not 0.5”). (3) Auxiliary GVFs stay on their own gamma: head 1 matches across `terminated in {0, 1}` (`-0.18965168` both sides). (4) SARSA parity: keep a `terminated=1` SARSA smoke so the documented flag is not `horde_ac`-only. (5) Assert on `AlbertaPipelineStepResult.state.control_state`, not a reconstructed discount scalar.

Until `pipeline.update(..., terminated=1.0)` on this recipe yields **empty actor traces** (issue origin matrix → zeros) **and** continuing value-head `0.34537804` with aux head 1 unchanged, this does not lock “forward termination-zeroed discount to HordeActorCriticAgent in horde_ac mode” — it locks a single `isclose(..., 0.5)` on `horde_td_targets[0]`, and the comment about traces asserts a different array.

**Review by ss251 — CHANGES_REQUESTED** [review]
FIX — exact head `6b0c0ebd520eba4c54b45a57b4b48a43e2bbe60e`.

The terminal repair itself is correct. In a fresh credential-free, network-disabled container, the focused regression passed:

```text
1 passed in 7.53s
```

The public `AlbertaPipeline.update` path produced the claimed terminal behavior:

```text
terminated 0.0  td_targets [ 0.34537804 -0.18965168]
terminated 1.0  td_targets [ 0.5        -0.18965168]
terminal actor_trace_weights [[0. 0. 0.] [0. 0. 0.]]
```

Blocking finding:

- **Preserve the legacy nonterminal kernel** (`alberta_framework/pipeline.py:1576`). The patch passes the configured value gamma explicitly on every nonterminal step. That switches `HordeActorCriticAgent.update` from the legacy omitted-`discount` kernel to `update_with_discounts`. The two are mathematically equivalent but not bit-identical for all valid states, so this changes continuing learning trajectories even when `terminated == 0`.

An exact base/head comparison from the same public pipeline state reproduced the drift:

```text
base targets [-0.04676366 -0.89790446]
head targets [-0.04676366 -0.8979045 ]
base tree_sha256 43d883c4bc63070164ad878f74656721a2f635533676421022a40ff63c985430
head tree_sha256 82ffdb8ae5714f9898c387ec1c1d3c762b448708c60397b8ed50874b59b02102
```

A direct same-state scan found eight differing result/state leaves in that configuration, with maximum absolute difference `5.960464477539063e-08`, and reproduced drift across multiple valid architectures. In a long-lived online learner, this perturbs every nonterminal trajectory and contradicts the PR's no-change reduction premise.

Please keep `terminated == 0` on the original omitted-`discount` call, use explicit zero discount only for terminal transitions, and add an exact full-tree base/head nonterminal equality regression. Retain the verified terminal target and actor-trace assertions.

Recommendation: **FIX**.

AI provider/model: OpenAI / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [assigned-asi-pr-review]

**Comment by lalalune** [comment]
Reviewed: Superseded by #2348 after comparing diffs: same terminal value discount correction, with fuller trace and auxiliary-head coverage in the retained implementation.
